### PR TITLE
Homework in ./lab/homework

### PR DESCRIPTION
**Pull request by alexandra-udrescu**

### CODE REVIEW

File: lab/homework/index.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+export class Customer extends Person {
+    
+    constructor({money, shoppingCart}) {
+        super({});
+        this.money = money;
+        this.shoppingCart = shoppingCart;
+    }
+    
+    addToCart(product) {
+        if (product instanceof Product) {
```
Comment (adrian-branescu) [inline lab/homework/index.js:84]: It's recommended to avoid nested `if` statements if that's possible.
In this case, an equivalent alternative, more readable and less error prone would be to test negative scenarios first:
```javascript
if (!(product instanceof Product)) {
    throw new HomeworkError({ code: HomeworkError.ERROR_CODE.NOT_A_PRODUCT });
}

if ((product instanceof Aliment) && product.isExpired()) {
    throw new HomeworkError({ code: HomeworkError.ERROR_CODE. EXPIRED });
}

/**
 *  if the execution reaches this point, it means:
 *    1. the product is a Product instance
 *    2. if it's also an Aliment instance then for sure it's not expired
 */
this.shoppingCart.push(product);
```

File: lab/homework/index.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            if(product instanceof Aliment && product.isExpired()) {
```
Comment (adrian-branescu) [inline lab/homework/index.js:85]: ~~Note that in the previous comment I used the following branch condition:~~
```javascript
if ((product instanceof Aliment) && product.isExpired()) { }
```
~~Take care of the operator precedence. In this case, `&&` is evaluated before `instanceof` operator if we don't use brackets.~~

Take a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table

File: lab/homework/index.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            throw new HomeworkError({
+                code: HomeworkError.ERROR_CODE.NOT_A_PRODUCT
+            });
+        }
+    }
+    
+    removeFromCart(product) {
+        if (product instanceof Product) {
+            // find element in shopping cart
+            var idx = this.shoppingCart.findIndex((el) => el.toString() === product.toString())
```
Comment (adrian-branescu) [inline lab/homework/index.js:102]: There are two problems here:

1. Never use `var` in your code. Use only `let` or `const`. 
In this case `const` is more appropriate since you don't plan to change its value.

2. When you call `.toString()` method on an instance of a class that doesn't have this method implemented it will use the one from the `Object` (implicit parent) class which returns the primitive string `[object Object]`.
In your case all your products will appear to be equal, so you will always remove the first one from that Array. 
For simplicity, you could directly use `===` to compare object references.

File: lab/homework/index.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    
+    scanAndCash(customer) {
+        let total = 0;
+        // compute total amount to pay
+        for (let prod of customer.shoppingCart) {
+            total = total + prod.price;
+        }
+        // check if the customer affords the cart
+        if(customer.money >= total) {
+            customer.money = customer.money - total;
```
Comment (adrian-branescu) [inline lab/homework/index.js:137]: `shoppingCart` and `money` properties should be private and you should only expose them via getters as read-only properties, otherwise any other entity can play with `Customer` internals.

If you want to manipulate them you should implement methods in the `Customer` class and call them where needed.

File: lab/homework/test.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+
+describe('Customer ERRORS', () => {
+
+    it('a customer won\'t add to his shopping cart an aliment that is expired', () => {
+        let customer1 = new Customer({money: 100, shoppingCart: new Array()});
+        let prod0 = new Furniture({name: "sofa", price: 70, colour: "blue"});
+        let prod1 = new Aliment({name: "pudding", price: 10, expirationDate: new Date() + 1, grammage: 50});
+        let prod2 = new Aliment({name: "puddingx3", price: 30, expirationDate: new Date() + 1, grammage: 150});
+        let prod3 = new Aliment({name: "puddingx3", price: 30, expirationDate: new Date() - 7, grammage: 150});
```
Comment (adrian-branescu) [inline lab/homework/test.js:15]: Take care that:

1. `new Date() - 7` converts the Date instance to number, namely how many milliseconds passed from 01.01.1970 until present (UNIX timestamps) and then performs arithmetic.

2.  `new Date() + 1` converts the Date instance to its string representation and then concatenates it with the other operand.

You should use `Date.now()` if you want to operate on UNIX timestamps and make sure arithmetic works fine.

File: lab/homework/package.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "13-npm",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha -bdd test.js"
+  },
+  "dependencies": {
```
Comment (adrian-branescu) [inline lab/homework/package.json:9]: These are actually `devDependencies` as you use them in tests, not in the source code of your package.
